File: app/src/main/java/vedam/subkuch/ui/chat/ChatListFragment.java
package vedam.subkuch.ui.chat;

import android.content.Intent;
import androidx.databinding.DataBindingUtil;
import android.os.Bundle;
import androidx.annotation.NonNull;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.android.volley.Response;

import java.util.ArrayList;

import vedam.subkuch.R;
import vedam.subkuch.base.BaseFragment;
import vedam.subkuch.databinding.FragmentChatListBinding;
import vedam.subkuch.helpers.Constants;
import vedam.subkuch.interfaces.OnListViewItemClickListener;
import vedam.subkuch.network.DataFetcher;
import vedam.subkuch.ui.matrimonial.models.DatingProfile;
import vedam.subkuch.ui.matrimonial.models.DatingProfileResponse;
import vedam.subkuch.utils.AppUtil;
import vedam.subkuch.utils.ListItemClickAction;
import vedam.subkuch.utils.UiUtil;


public class ChatListFragment extends BaseFragment implements OnListViewItemClickListener {
    private FragmentChatListBinding fragmentChatListBinding;
    private ChatListAdapter adapter;
    private ArrayList<DatingProfile> datingProfiles = new ArrayList<>();
    private boolean loading = true;
    private LinearLayoutManager linearLayoutManager;
    private int pageNo = 1;
    private int pageSize = 20;
    private boolean hasMoreProjects = true;
    private boolean isDating;

    public ChatListFragment() {
        // Required empty public constructor
    }


    public static ChatListFragment newInstance(boolean isDating) {
        ChatListFragment chatListFragment = new ChatListFragment();
        Bundle bundle = new Bundle();
        bundle.putBoolean(Constants.EXTRA_IS_DATING, isDating);
        chatListFragment.setArguments(bundle);
        return chatListFragment;
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        if (getArguments() != null)
            isDating = getArguments().getBoolean(Constants.EXTRA_IS_DATING);
    }

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        // Inflate the layout for this fragment
        fragmentChatListBinding = DataBindingUtil.inflate(inflater, R.layout.fragment_chat_list, container, false);
        return fragmentChatListBinding.getRoot();
    }

    public void onViewCreated(@NonNull View v, Bundle savedInstanceState) {
        super.onViewCreated(v, savedInstanceState);
        setTitle(getString(R.string.matches));
        initUI();
        getMatchedProfiles();
    }

    private void initUI() {

        linearLayoutManager = new LinearLayoutManager(context);
        fragmentChatListBinding.rvChatList.setLayoutManager(linearLayoutManager);
        fragmentChatListBinding.rvChatList.setHasFixedSize(true);
        adapter = new ChatListAdapter(context, datingProfiles, this);
        fragmentChatListBinding.rvChatList.setAdapter(adapter);
        fragmentChatListBinding.rvChatList.addOnScrollListener(new ProfilesOnScrollListener());

    }

    public void getMatchedProfiles() {
        UiUtil.showProgressDialog(context, getString(R.string.please_wait));
        if (isDating)
            DataFetcher.getDatingMatchedChatProfiles(context, onMatchedProfilesSuccessListener, DatingProfileResponse.class, onErrorListener, pageNo, pageSize);
        else
            DataFetcher.getMatrimonialMatchedChatProfiles(context, onMatchedProfilesSuccessListener, DatingProfileResponse.class, onErrorListener, pageNo, pageSize);

    }

    private void startChatActivity(DatingProfile datingProfile) {

        Intent intent = new Intent(context, ChatActivity.class);
        intent.putExtra(Constants.EXTRA_NAME, AppUtil.deNull(datingProfile.getFirstName()));
        intent.putExtra(Constants.EXTRA_CHAT_TO_ID, datingProfile.getProfileId());
        intent.putExtra(Constants.EXTRA_IS_DATING, isDating);
        startActivityForResult(intent, Constants.REQUEST_CHAT);
    }

    private Response.Listener<DatingProfileResponse> onMatchedProfilesSuccessListener = response -> {

        UiUtil.cancelProgressDialog();
        if (getActivity() != null)
            if (response != null && response.getReturnMessage().equals(Constants.SUCCESS)) {
                if (response.getReturnData().size() > 0) {
                    hasMoreProjects = response.getReturnData().size() >= pageSize;
                    loading = true;
                    loadValues(response.getReturnData());
                } else
                    UiUtil.showToast(context, getString(R.string.no_matches_found));
            } else
                UiUtil.showToast(context, getString(R.string.err_occurred));
    };

    private void loadValues(ArrayList<DatingProfile> response) {

        if (response != null && !response.isEmpty()) {
            pageNo++;
            datingProfiles.addAll(response);
            adapter = new ChatListAdapter(context, datingProfiles, this);
            fragmentChatListBinding.rvChatList.setAdapter(adapter);
        }
    }

    @Override
    public <E> void onItemClick(E item, int position, View view, ListItemClickAction action) {
        if (item != null) {
            DatingProfile datingProfile = (DatingProfile) item;
            startChatActivity(datingProfile);
        }
    }

    public class ProfilesOnScrollListener extends RecyclerView.OnScrollListener {

        @Override
        public void onScrollStateChanged(@NonNull RecyclerView recyclerView, int newState) {
            super.onScrollStateChanged(recyclerView, newState);
            /*final Picasso picasso = Picasso.get();

            if (newState == RecyclerView.SCROLL_STATE_IDLE || newState == RecyclerView.SCROLL_STATE_DRAGGING) {
                picasso.resumeTag(context);
            } else {
                picasso.pauseTag(context);
            }*/
        }

        @Override
        public void onScrolled(@NonNull RecyclerView recyclerView, int dx, int dy) {
            if (dy > 0) //check for scroll down
            {
                int visibleItemCount = linearLayoutManager.getChildCount();
                int totalItemCount = linearLayoutManager.getItemCount();
                int pastVisibleItems = linearLayoutManager.findFirstVisibleItemPosition();

                if (loading) {
                    if ((visibleItemCount + pastVisibleItems) >= totalItemCount) {
                        loading = false;
                        if (hasMoreProjects) getMatchedProfiles();

                    }
                }
            }
        }
    }

    @Override
    public void onActivityResult(int requestCode, int resultCode, Intent data) {
        if (requestCode == Constants.REQUEST_CHAT) {
            refreshData();
        } else {
            super.onActivityResult(requestCode, resultCode, data);
        }
    }

    private void refreshData() {
        pageNo = 1;
        hasMoreProjects = true;
        loading = true;
        datingProfiles.clear();
        getMatchedProfiles();
    }

    public void changeData() {
        adapter.notifyDataSetChanged();
    }

}

File: app/src/main/java/vedam/subkuch/ui/chat/ChatListFragment.kt
package vedam.subkuch.ui.chat

import android.content.Intent
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Toast
import androidx.databinding.DataBindingUtil
import androidx.recyclerview.widget.LinearLayoutManager
import vedam.subkuch.network.Response
import com.google.android.gms.tasks.Tasks
import com.google.firebase.firestore.DocumentSnapshot
import com.google.firebase.firestore.FirebaseFirestore
import com.google.firebase.firestore.QuerySnapshot
import vedam.subkuch.R
import vedam.subkuch.base.BaseFragment
import vedam.subkuch.databinding.FragmentChatListBinding
import vedam.subkuch.db.chat.LatestChat
import vedam.subkuch.helpers.Constants
import vedam.subkuch.interfaces.OnListViewItemClickListener
import vedam.subkuch.network.DataFetcher
import vedam.subkuch.ui.matrimonial.models.DatingProfile
import vedam.subkuch.ui.matrimonial.models.DatingProfileResponse
import vedam.subkuch.utils.AppPrefs
import vedam.subkuch.utils.AppUtil
import vedam.subkuch.utils.ListItemClickAction
import vedam.subkuch.utils.UiUtil


class ChatListFragment : BaseFragment(), OnListViewItemClickListener {
    private var fragmentChatListBinding: FragmentChatListBinding? = null
    private var adapter: ChatListAdapter? = null
    private var datingProfiles: List<DatingProfile>? = null
    private var loading = true

    //    private var pageNo = 1
    private val pageSize = 200
    private var hasMoreProjects = true
    private var isDating = false
    private val latestChatMap = mutableMapOf<String, LatestChat>()
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        isDating = arguments?.getBoolean(Constants.EXTRA_IS_DATING) ?: false
    }

    override fun onCreateView(
        inflater: LayoutInflater, container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        // Inflate the layout for this fragment
        fragmentChatListBinding =
            DataBindingUtil.inflate(inflater, R.layout.fragment_chat_list, container, false)
        return fragmentChatListBinding?.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        setTitle(getString(R.string.matches))
        initUI()
    }

    override fun onResume() {
        super.onResume()
        getMatchedProfiles()
    }

    private fun getLatestChats() {

        val task1 = FirebaseFirestore.getInstance().collection(Constants.TABLE_LATEST_CHAT)
            .whereEqualTo(Constants.ToProfileId, AppPrefs.getPrefsUserId(mContext)).get()

        val task2 = FirebaseFirestore.getInstance().collection(Constants.TABLE_LATEST_CHAT)
            .whereEqualTo(Constants.FromProfileId, AppPrefs.getPrefsUserId(mContext)).get()

        Tasks.whenAllSuccess<QuerySnapshot>(task1, task2).addOnSuccessListener {
            for (snapshot in it)
                for (doc in snapshot.documents) {
                    val chat = getLatestChatFromDoc(doc)
                    latestChatMap[chat.idPair!!] = getLatestChatFromDoc(doc)
                }
            filterChatList()
        }.addOnFailureListener {
            UiUtil.cancelProgressDialog()
        }
    }

    private fun filterChatList() {

        val myId = AppPrefs.getPrefsUserId(mContext)
        datingProfiles?.let { list ->
            for (profile in list) {
                val idPair = myId.getIdPair(profile.ProfileId!!)
                val latestChat = latestChatMap[idPair]
                latestChat?.let { profile.latestChat = it }
            }
            val chatList = list.filter { it.latestChat != null }
                .sortedByDescending { it.latestChat?.timeStamp }
            if (chatList.isNotEmpty())
                adapter?.submitList(chatList)
            else
                Toast.makeText(
                    mContext,
                    "No Chats. Go to your Matches to start a Chat.",
                    Toast.LENGTH_LONG
                ).show()
            UiUtil.cancelProgressDialog()
        } ?: kotlin.run {
            UiUtil.cancelProgressDialog()
        }
    }

    private fun getLatestChatFromDoc(doc: DocumentSnapshot): LatestChat {

        val latestChat = LatestChat()
        latestChat.id = doc.id
        latestChat.fromProfileId = doc.getString("fromProfileId")
        latestChat.toProfileId = doc.getString("toProfileId")
        latestChat.senderName = doc.getString("senderName")
        latestChat.latestMessage = doc.getString("latestMessage")
        latestChat.timeStamp =
            doc.getTimestamp("timeStamp")
        latestChat.idPair = doc.getString(Constants.idPair)

        return latestChat
    }

    private fun initUI() {
        fragmentChatListBinding!!.rvChatList.layoutManager = LinearLayoutManager(mContext)
        adapter = ChatListAdapter(requireContext(), this)
        fragmentChatListBinding!!.rvChatList.adapter = adapter

    }

    private fun String.getIdPair(chatToId: String): String {
        return if (this > chatToId)
            "${this}_$chatToId"
        else
            "${chatToId}_${this}"
    }

    private fun getMatchedProfiles() {
        UiUtil.showProgressDialog(mContext, getString(R.string.please_wait))
        if (isDating) DataFetcher.getDatingMatchedChatProfiles(
            mContext,
            onMatchedProfilesSuccessListener,
            DatingProfileResponse::class.java,
            onErrorListener,
            1,
            pageSize
        ) else DataFetcher.getMatrimonialMatchedChatProfiles(
            mContext,
            onMatchedProfilesSuccessListener,
            DatingProfileResponse::class.java,
            onErrorListener,
            1,
            pageSize
        )
    }

    private fun startChatActivity(datingProfile: DatingProfile) {
        val intent = Intent(mContext, ChatActivity::class.java)
        intent.putExtra(Constants.EXTRA_NAME, AppUtil.deNull(datingProfile.FirstName))
        intent.putExtra(Constants.EXTRA_CHAT_TO_ID, datingProfile.ProfileId)
        intent.putExtra(Constants.EXTRA_IS_DATING, isDating)
        startActivity(intent)
    }

    private val onMatchedProfilesSuccessListener =
        Response.Listener { response: DatingProfileResponse? ->
            if (activity != null) if (response != null && response.returnMessage == Constants.SUCCESS) {
                if (response.returnData.size > 0) {
//                    hasMoreProjects = response.returnData.size >= pageSize
//                    loading = true
                    datingProfiles = response.returnData
                    getLatestChats()
                } else {
                    UiUtil.cancelProgressDialog()
                    UiUtil.showToast(mContext, getString(R.string.no_matches_found))
                }
            } else {
                UiUtil.cancelProgressDialog()
                UiUtil.showToast(mContext, getString(R.string.err_occurred))
            }
        }

    override fun <E> onItemClick(
        item: E?,
        position: Int,
        view: View?,
        action: ListItemClickAction?
    ) {
        if (item != null) {
            val datingProfile = item as DatingProfile
            startChatActivity(datingProfile)
        }
    }

    companion object {
        @JvmStatic
        fun newInstance(isDating: Boolean): ChatListFragment {
            val chatListFragment = ChatListFragment()
            val bundle = Bundle()
            bundle.putBoolean(Constants.EXTRA_IS_DATING, isDating)
            chatListFragment.arguments = bundle
            return chatListFragment
        }
    }
}
